File: python/ad_api.py
import ldap
import json
import os
from flask import Flask, render_template

app = Flask(__name__)

def validUser(auid):
    ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
    ldap.set_option(ldap.OPT_REFERRALS, 0)
    ldap.set_option(ldap.OPT_PROTOCOL_VERSION, 3)

    AD = ldap.initialize(os.environ['LDAP_URL'])

    base_dn = os.environ['BASE_DN']
    search_scope = ldap.SCOPE_SUBTREE
    search_filter = "(&(objectCategory=User)(|(sAMAccountName=" + auid + ")))"
    get_attributes = ['sAMAccountName']

    AD.simple_bind_s('', '')

    search = AD.search_s(base_dn, search_scope, search_filter, get_attributes)[0]

    if search[0] == None:
        return False
    else:
        return True


def userInfo(auid):
    ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
    ldap.set_option(ldap.OPT_REFERRALS, 0)
    ldap.set_option(ldap.OPT_PROTOCOL_VERSION, 3)

    AD = ldap.initialize(os.environ['LDAP_URL'])

    base_dn = os.environ['BASE_DN']
    search_scope = ldap.SCOPE_SUBTREE
    search_filter = "(&(objectCategory=User)(|(sAMAccountName=" + auid + ")))"
    get_attributes = ['sAMAccountName', 'mail', 'sn', 'givenName', 'initials', 'title']

    AD.simple_bind_s('', '')

    search = AD.search_s(base_dn, search_scope, search_filter, get_attributes)[0]

    result = dict(search[1])
    for key in result:
        result[key] = result[key][0].decode("utf-8")

    result["cn"] = search[0]

    result_json = json.dumps(result, indent = 4)
    # print(result_json)

    AD.unbind_s()

    return result_json


def userGroups(auid):
    ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
    ldap.set_option(ldap.OPT_REFERRALS, 0)
    ldap.set_option(ldap.OPT_PROTOCOL_VERSION, 3)

    AD = ldap.initialize(os.environ['LDAP_URL'])

    base_dn = os.environ['BASE_DN']
    search_scope = ldap.SCOPE_SUBTREE
    search_filter = "(&(objectCategory=User)(|(sAMAccountName=" + auid + ")))"
    get_attributes = ['dn']
    groups = []

    AD.simple_bind_s('', '')

    search = AD.search_s(base_dn, search_scope, search_filter, get_attributes)[0]

    fixed_cn = search[0].replace('\\', "\\\\")

    group_filter = "(member="+fixed_cn+")"

    group_search = AD.search_s(
        base_dn, search_scope, group_filter, ['cn', 'description'])

    for item in group_search:
        if isinstance(item[1], dict):
            group_obj = item[1]
            for key in group_obj:
                group_obj[key] = group_obj[key][0].decode("utf-8")

            groups.append(group_obj)



    AD.unbind_s()

    result_json = json.dumps(groups, indent = 4)
    # print(result_json)

    return result_json


def allInfo(auid):
    ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
    ldap.set_option(ldap.OPT_REFERRALS, 0)
    ldap.set_option(ldap.OPT_PROTOCOL_VERSION, 3)

    AD = ldap.initialize(os.environ['LDAP_URL'])
    base_dn = os.environ['BASE_DN']
    search_scope = ldap.SCOPE_SUBTREE
    search_filter = "(&(objectCategory=User)(|(sAMAccountName=" + auid + ")))"
    get_attributes = ['sAMAccountName', 'mail', 'sn', 'givenName', 'initials', 'title']
    groups = []
    all_info = {}

    AD.simple_bind_s('', '')

    search = AD.search_s(base_dn, search_scope, search_filter, get_attributes)[0]

    if search[0] == None:
        return "No user found"

    user_result = dict(search[1])
    for key in user_result:
        user_result[key] = user_result[key][0].decode("utf-8")

    all_info['valid'] = True

    all_info['user'] = user_result

    fixed_cn = search[0].replace('\\', "\\\\")

    group_filter = "(member="+fixed_cn+")"

    group_search = AD.search_s(base_dn, search_scope, group_filter, ['cn', 'description'])

    for item in group_search:
        if isinstance(item[1], dict):
            group_obj = item[1]
            for key in group_obj:
                group_obj[key] = group_obj[key][0].decode("utf-8")

            groups.append(group_obj)
    
    all_info['groups'] = groups

    AD.unbind_s()

    result_json = json.dumps(all_info, indent = 4)

    return result_json


@app.route("/ad_api/")
def index():
    current_user = os.environ['USERNAME']
    cu_info = json.loads(getAll(current_user))
    user = json.dumps(cu_info['user'], sort_keys = True, indent = 4, separators = (',', ': '))
    groups = json.dumps(cu_info['groups'], sort_keys = True, indent = 4, separators = (',', ': '))
    valid = cu_info['valid']
    # return render_template('index.html', data=cu_info)
    return render_template('index.html', data=cu_info, valid=valid, user=user, groups=groups)

@app.route("/ad_api/validuser/<auid>")
def getValid(auid):
    isValid = validUser(auid)
    return str(isValid)

@app.route("/ad_api/userinfo/<auid>")
def getUser(auid):
    user = userInfo(auid)
    return user

@app.route("/ad_api/usergroups/<auid>")
def getGroups(auid):
    groups = userGroups(auid)
    return groups

@app.route("/ad_api/all/<auid>")
def getAll(auid):
    allinfo = allInfo(auid)
    return allinfo


if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True, port=8080)

File: js/ad_api.js
const path = require('path')
const ldap = require('ldapjs');
const fastify = require('fastify')({ logger: false })
const helmet = require('@fastify/helmet')
// const pino = require('pino')
// const logger = pino({
//     transport: {
//       target: 'pino-pretty'
//     },
//     level: 'trace'
// })

const client = ldap.createClient({
  url: 'process.env.LDAP_URL',
  tlsOptions: {rejectUnauthorized: false},
  reconnect: true,
  // timeout: 1000,
  // idleTimeout: 1000
})
client.on('connect', (connection) => {
  // console.log("LDAP Connection Success");
  // console.log(connection);
})
client.on('timeout', (err) => {
  console.log("timed out: " + err)
})
client.on('error', (err) => {
  // console.log(err);
})
client.on('destroy', () => {
  console.log("ldap client destoryed");
})

const validUser = (auid, callback) => {
  client.bind('', '', (err) => {
    if(err) {
      callback(err)
      return
    }

    const opts = {
      filter: `(&(|(agencyUID=${auid})))`,
      scope: 'sub',
      attributes: ['agencyUID']
    };
  
    client.search(process.env.BASE_DN, opts, (err, res) => {
      let user = ""
  
      res.on('searchEntry', (entry) => {
        user = JSON.stringify(entry.object)
        // console.log(JSON.stringify(entry));
        // callback(JSON.stringify(entry.object))
      });
  
      res.on('error', (err) => {
        console.error('error: ' + err.message);
      });
  
      res.on('end', (result) => {
        if(user == ""){
          user = false
          callback(user)
        }
        else{
          user = true
          callback(user)
        }
        // client.destroy()
        client.unbind((err) => {})
      });
    })
  });
}

const findUser = (auid, callback) => {
  client.bind('', '', (err) => {
    if(err) {
      callback(err)
      return
    }

    const opts = {
      filter: `(&(|(agencyUID=${auid})))`,
      scope: 'sub',
      attributes: ['agencyUID', 'mail', 'sn', 'givenName', 'initials', 'title']
    };
  
    client.search(process.env.BASE_DN, opts, (err, res) => {
      let user = ""
  
      res.on('searchEntry', (entry) => {
        user = entry.object
      });
  
      res.on('error', (err) => {
        console.error('error: ' + err.message);
      });
  
      res.on('end', (result) => {
        if(user == ""){
          user = "No user found."
        }
        callback(user)
        // client.destroy()
        client.unbind((err) => {})
      });
    })
  });
}

const findUserGroups = (auid, callback) => {
  client.bind('', '', (err) => {
    if(err) {
      callback(err)
      return
    }
    const opts = {
      filter: `(&(|(agencyUID=${auid})))`,
      scope: 'sub',
      attributes: ['agencyUID', 'mail', 'sn', 'givenName', 'initials', 'title']
    };
    client.search(process.env.BASE_DN, opts, (err, res) => {
      let user = ""
  
      res.on('searchEntry', (entry) => {
        user = entry.object
      });
  
      res.on('error', (err) => {
        console.error('error: ' + err.message);
      });
  
      res.on('end', (result) => {
        if(user == ""){
          user = {}
        }
        groupSearch(user)
      });
    })
    const groupSearch = (userObj) => {
      // const dn = userObj.dn.replace("\\", "\\\\")
      const dn = "NDC\\\\dcrendon"
      const opts = {
        filter: `(member=${dn})`,
        scope: 'sub',
      }
      client.search(process.env.BASE_DN, opts, (err, res) => {
        let groups = []
        res.on('searchEntry', (entry) => {
          const group = entry.object
          groups.push(group)
        });
        res.on('error', (err) => {
          console.error('error: ' + err.message);
        });
        res.on('end', (result) => {
          if(groups.length == 0){
            groups = "No groups found"
          }
          callback(groups)
          client.unbind((err) => {})
        });
      })
    }
  })
}

const allInfo = (auid, callback) => {
  client.bind('', '', (err) => {
    if(err) {
      callback(err)
      return
    }
    const opts = {
      filter: `(&(|(agencyUID=${auid})))`,
      scope: 'sub',
      attributes: ['agencyUID']
    };
    client.search(process.env.BASE_DN, opts, (err, res) => {
      let user = ""
      res.on('searchEntry', (entry) => {
        user = JSON.stringify(entry.object)
      });
      res.on('error', (err) => {
        console.error('error: ' + err.message);
      });
  
      res.on('end', (result) => {
        if(user == ""){
          user = {}
          callback(user)
        }
        else{
          user = true
          findUser(auid)
        }
      });
    })
    const findUser = (auid) => {
      const opts = {
        filter: `(&(|(agencyUID=${auid})))`,
        scope: 'sub',
        attributes: ['agencyUID', 'mail', 'sn', 'givenName', 'initials', 'title']
      };
      client.search("dc=nasa,dc=gov", opts, (err, res) => {
        let user = ""
    
        res.on('searchEntry', (entry) => {
          user = entry.object
        });
    
        res.on('error', (err) => {
          console.error('error: ' + err.message);
        });
    
        res.on('end', (result) => {
          if(user == ""){
            user = {}
          }
          groupSearch(user)
        });
      })
    }
    const groupSearch = (userObj) => {
      const dn = userObj.dn.replace("\\", "\\\\")
      const opts = {
        filter: `(member=${dn})`,
        scope: 'sub',
        attributes: ['cn', 'description']
      }
      client.search("dc=nasa,dc=gov", opts, (err, res) => {
        let groups = []
        res.on('searchEntry', (entry) => {
          const group = entry.object
          groups.push(group)
        });
        res.on('error', (err) => {
          console.error('error: ' + err.message);
        });
        res.on('end', (result) => {
          if(groups.length == 0){
            groups = "No groups found"
          }
          const allInfo = {
            isValidUser: true,
            user: userObj,
            groups: groups
          }
          callback(allInfo)
          client.unbind((err) => {})
        });
      })
    }
  })
}

fastify.register(helmet, { global: true })

fastify.register(require("@fastify/view"), {
  engine: {
    eta: require("eta"),
  },
})

fastify.register(require('@fastify/static'), {
    root: path.join(__dirname, 'public'),
    prefix: '/ad_api/public/',
})

fastify.get('/ad_api/', (request, reply) => {
  allInfo("dcrendon", (allInfo) => {
    if(allInfo.error) reply.send(allInfo)
    reply.view("index.eta", allInfo)
  })
})

fastify.get('/ad_api/userinfo/:auid', (request, reply) => {
    const { auid } = request.params
    findUser(auid, (user) => {
      reply.send(user)
    })
})

fastify.get('/ad_api/usergroups/:auid', (request, reply) => {
    const { auid } = request.params
    findUserGroups(auid, (groups) => {
      reply.send(groups)
    })
})

fastify.get('/ad_api/validuser/:auid', (request, reply) => {
    const { auid } = request.params
    validUser(auid, (isValid) => {
      reply.send(isValid)
    })
})

fastify.get('/ad_api/all/:auid', (request, reply) => {
    const { auid } = request.params
    allInfo(auid, (allInfo) => {
      reply.send(allInfo)
    })
})

fastify.listen({port: process.env.PORT.trim()})